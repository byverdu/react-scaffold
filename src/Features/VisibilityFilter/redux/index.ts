import { createAction } from 'redux-actions';
import { VisibilityFilterEnum } from 'Models/Enums';
import { VisibilityFilterState } from 'Models/Todo';

export const SET_ACTIVE_FILTER = 'SET_ACTIVE_FILTER';
export const setActiveFilter = createAction<VisibilityFilterEnum>(
  SET_ACTIVE_FILTER
);

export const initialState: VisibilityFilterState = {
  activeFilter: VisibilityFilterEnum.SHOW_ALL
};

const visibilityFilter = (
  state = initialState,
  action
): VisibilityFilterState => {
  switch (action.type) {
    case SET_ACTIVE_FILTER:
      return {
        activeFilter: action.payload
      };
    default:
      return state;
  }
};

export default visibilityFilter;
