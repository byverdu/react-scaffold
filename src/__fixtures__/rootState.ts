import { TodoState, VisibilityFilterState } from 'Models/Todo';
import { LoggerState } from 'Models/Logger';
import { RootState } from 'Core/reducers';
import { VisibilityFilterEnum } from 'Models/Enums';
import { Map } from 'immutable';
import { mockedData } from '__fixtures__/actions';

const todos: TodoState = {
  todos: Map(mockedData)
};

const visibilityFilter: VisibilityFilterState = {
  activeFilter: VisibilityFilterEnum.SHOW_ALL
};

export const logger: LoggerState = {
  type: undefined,
  message: '',
  isVisible: false
};

const rootState: RootState = {
  todos,
  visibilityFilter,
  logger
};

export default rootState;
