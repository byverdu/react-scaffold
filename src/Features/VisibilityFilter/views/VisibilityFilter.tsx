import React from 'react';
import { connect } from 'react-redux';
import { VisibilityFilterEnum } from 'Models/Enums';
import { RootState } from 'Core/reducers';
import { setActiveFilter } from 'Features/VisibilityFilter/redux';

export interface FilterMapStateToProps {
  activeFilter: VisibilityFilterEnum;
}

export interface FilterDispatchToProps {
  filterHandler: (filter: VisibilityFilterEnum) => void;
}

interface ComponentProps extends FilterMapStateToProps, FilterDispatchToProps {}

const mapStateToProps = ({
  visibilityFilter: { activeFilter }
}: RootState): FilterMapStateToProps => ({
  activeFilter
});

const mapDispatchToProps = (dispatch): FilterDispatchToProps => ({
  filterHandler: (filter: VisibilityFilterEnum) =>
    dispatch(setActiveFilter(filter))
});

const VisibilityFilter: React.SFC<ComponentProps> = ({
  activeFilter,
  filterHandler
}) => {
  const filtersData = [
    {
      text: 'Show All',
      type: VisibilityFilterEnum.SHOW_ALL
    },
    {
      text: 'Show Done',
      type: VisibilityFilterEnum.SHOW_DONE
    },
    {
      text: 'Show Todo',
      type: VisibilityFilterEnum.SHOW_TODO
    }
  ].map((item) => (
    <label key={item.type} htmlFor={item.type}>
      {item.text}
      <input
        checked={activeFilter === item.type}
        name="filter-type"
        id={item.type}
        value={item.type}
        type="radio"
        onChange={() => {
          filterHandler(item.type);
        }}
      />
    </label>
  ));

  return <section className="visibility-actions">{filtersData}</section>;
};

export default connect<FilterMapStateToProps, FilterDispatchToProps>(
  mapStateToProps,
  mapDispatchToProps
)(VisibilityFilter);
