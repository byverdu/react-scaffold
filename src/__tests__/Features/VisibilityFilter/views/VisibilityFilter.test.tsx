import React from 'react';
import ReactDOM from 'react-dom';
import VisibilityFilter from 'Features/VisibilityFilter/views/VisibilityFilter';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import rootState from '__fixtures__/rootState';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore(rootState);

describe('<VisibilityFilter />', () => {
  let wrapper;
  const component = () => (
    <Provider store={store}>
      <VisibilityFilter />
    </Provider>
  );

  beforeEach(() => {
    wrapper = mount(component());
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component(), div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render 3 inputs', () => {
    expect(wrapper.find('input[type="radio"]')).toHaveLength(3);
  });

  it('SHOW_ALL input should be checked by default', () => {
    expect(wrapper.find('input[id="SHOW_ALL"]').props().checked).toEqual(true);
  });
});
