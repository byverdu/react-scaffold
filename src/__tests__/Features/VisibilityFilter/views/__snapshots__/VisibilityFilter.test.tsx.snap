// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisibilityFilter /> should create a snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(VisibilityFilter)>
    <VisibilityFilter
      activeFilter="SHOW_ALL"
      filterHandler={[Function]}
    >
      <section>
        <label
          htmlFor="SHOW_ALL"
          key="SHOW_ALL"
        >
          Show All
          <input
            checked={true}
            id="SHOW_ALL"
            name="filter-type"
            onChange={[Function]}
            type="radio"
            value="SHOW_ALL"
          />
        </label>
        <label
          htmlFor="SHOW_DONE"
          key="SHOW_DONE"
        >
          Show Done
          <input
            checked={false}
            id="SHOW_DONE"
            name="filter-type"
            onChange={[Function]}
            type="radio"
            value="SHOW_DONE"
          />
        </label>
        <label
          htmlFor="SHOW_TODO"
          key="SHOW_TODO"
        >
          Show Todo
          <input
            checked={false}
            id="SHOW_TODO"
            name="filter-type"
            onChange={[Function]}
            type="radio"
            value="SHOW_TODO"
          />
        </label>
      </section>
    </VisibilityFilter>
  </Connect(VisibilityFilter)>
</Provider>
`;
